File: clsZippedImagingFilesScanner.vb
Option Strict On

' Written by Matthew Monroe for the Department of Energy (PNNL, Richland, WA)
'
' Last modified July 23, 2012

Public Class clsZippedImagingFilesScanner

	Inherits clsMSFileInfoProcessorBaseClass

	Public Const ZIPPED_IMAGING_FILE_SEARCH_SPEC As String = "0_R*.zip"
	Public Const ZIPPED_IMAGING_FILE_NAME_PREFIX As String = "0_R"

	''' <summary>
	''' Examines the subdirectories in the specified zip file
	''' Determines the oldest and newest modified analysis.baf files (or apexAcquisition.method file if analysis.baf files are not found)
	''' Presumes this is the AcqStartTime and AcqEndTime
	''' </summary>
	''' <param name="ioZipFile"></param>
	''' <param name="udtFileInfo"></param>
	''' <returns>True if at least one valid file is found; otherwise false</returns>
	''' <remarks></remarks>
	Protected Function DetermineAcqStartEndTime(ByVal ioZipFile As System.IO.FileInfo, _
	   ByRef udtFileInfo As iMSFileInfoProcessor.udtFileInfoType) As Boolean

		Dim blnSuccess As Boolean = False

		Dim lstFileNamesToFind As System.Collections.Generic.List(Of String)
		Dim strNameParts() As String

		Try
			' Bump up the file size
			udtFileInfo.FileSizeBytes += ioZipFile.Length

			lstFileNamesToFind = New System.Collections.Generic.List(Of String)
			lstFileNamesToFind.Add("analysis.baf")
			lstFileNamesToFind.Add("apexAcquisition.method")
			lstFileNamesToFind.Add("submethods.xml")

			Dim oZipFile As Ionic.Zip.ZipFile
			Dim oZipEntry As System.Collections.Generic.IEnumerator(Of Ionic.Zip.ZipEntry)
			oZipFile = New Ionic.Zip.ZipFile(ioZipFile.FullName)

			For Each strFileNameToFind As String In lstFileNamesToFind
				oZipEntry = oZipFile.GetEnumerator()

				Do While oZipEntry.MoveNext

					If Not oZipEntry.Current.IsDirectory Then

						' Split the filename on the forward slash character
						strNameParts = oZipEntry.Current.FileName.Split("/"c)

						If Not strNameParts Is Nothing AndAlso strNameParts.Length > 0 Then

							If strNameParts(strNameParts.Length - 1).ToLower() = strFileNameToFind.ToLower() Then
								If oZipEntry.Current.LastModified < udtFileInfo.AcqTimeStart Then
									udtFileInfo.AcqTimeStart = oZipEntry.Current.LastModified
								End If

								If oZipEntry.Current.LastModified > udtFileInfo.AcqTimeEnd Then
									udtFileInfo.AcqTimeEnd = oZipEntry.Current.LastModified
								End If

								' Bump up the scan count
								udtFileInfo.ScanCount += 1

								' Add a Scan Stats entry
								Dim objScanStatsEntry As New DSSummarizer.clsScanStatsEntry

								objScanStatsEntry.ScanNumber = udtFileInfo.ScanCount
								objScanStatsEntry.ScanType = 1

								objScanStatsEntry.ScanTypeName = "MALDI-HMS"
								objScanStatsEntry.ScanFilterText = ""

								objScanStatsEntry.ElutionTime = "0"
								objScanStatsEntry.TotalIonIntensity = "0"
								objScanStatsEntry.BasePeakIntensity = "0"
								objScanStatsEntry.BasePeakMZ = "0"

								' Base peak signal to noise ratio
								objScanStatsEntry.BasePeakSignalToNoiseRatio = "0"

								objScanStatsEntry.IonCount = 0
								objScanStatsEntry.IonCountRaw = 0

								mDatasetStatsSummarizer.AddDatasetScan(objScanStatsEntry)


								blnSuccess = True

							End If
						End If

					End If
				Loop

				If blnSuccess Then Exit For
			Next

		Catch ex As System.Exception
			ReportError("Error finding XMass method folder: " & ex.Message)
			blnSuccess = False
		End Try

		Return blnSuccess

	End Function

	Protected Function GetDatasetFolder(ByVal strDataFilePath As String) As System.IO.DirectoryInfo

		Dim ioDatasetFolder As System.IO.DirectoryInfo
		Dim ioFileInfo As System.IO.FileInfo

		' First see if strFileOrFolderPath points to a valid file
		ioFileInfo = New System.IO.FileInfo(strDataFilePath)

		If ioFileInfo.Exists() Then
			' User specified a file; assume the parent folder of this file is the dataset folder
			ioDatasetFolder = ioFileInfo.Directory
		Else
			' Assume this is the path to the dataset folder
			ioDatasetFolder = New System.IO.DirectoryInfo(strDataFilePath)
		End If

		Return ioDatasetFolder

	End Function

	Public Overrides Function GetDatasetNameViaPath(ByVal strDataFilePath As String) As String
		Dim ioDatasetFolder As System.IO.DirectoryInfo
		Dim strDatasetName As String = String.Empty

		Try
			' The dataset name for a dataset with zipped imaging files is the name of the parent directory
			' However, strDataFilePath could be a file or a folder path, so use GetDatasetFolder to get the dataset folder
			ioDatasetFolder = GetDatasetFolder(strDataFilePath)
			strDatasetName = ioDatasetFolder.Name

			If strDatasetName.ToLower().EndsWith(".d") Then
				strDatasetName = strDatasetName.Substring(0, strDatasetName.Length - 2)
			End If

		Catch ex As System.Exception
			' Ignore errors
		End Try

		If strDatasetName Is Nothing Then strDatasetName = String.Empty
		Return strDatasetName

	End Function

	Public Shared Function IsZippedImagingFile(ByVal strFileName As String) As Boolean

		Dim ioFileInfo As System.IO.FileInfo = New System.IO.FileInfo(strFileName)

		If String.IsNullOrWhiteSpace(strFileName) Then
			Return False
		End If

		If ioFileInfo.Name.ToLower().StartsWith(ZIPPED_IMAGING_FILE_NAME_PREFIX.ToLower()) AndAlso ioFileInfo.Extension.ToLower() = ".zip" Then
			Return True
		Else
			Return False
		End If

	End Function

	Public Overrides Function ProcessDatafile(ByVal strDataFilePath As String, ByRef udtFileInfo As iMSFileInfoProcessor.udtFileInfoType) As Boolean
		' Process a Bruker Xmass folder, specified by strDataFilePath (which can either point to the dataset folder containing the XMass files, or any of the Zip files in the dataset folder)

		Dim ioDatasetFolder As System.IO.DirectoryInfo

		Dim ioFiles() As System.IO.FileInfo

		Dim blnSuccess As Boolean

		Try
			' Determine whether strDataFilePath points to a file or a folder

			ioDatasetFolder = GetDatasetFolder(strDataFilePath)

			' Validate that we have selected a valid folder
			If Not ioDatasetFolder.Exists Then
				MyBase.ReportError("File/folder not found: " & strDataFilePath)
				Return False
			End If

			' In case we cannot find any .Zip files, update the .AcqTime values to the folder creation date
			With udtFileInfo
				.AcqTimeStart = ioDatasetFolder.CreationTime
				.AcqTimeEnd = ioDatasetFolder.CreationTime
			End With

			' Look for the 0_R*.zip files
			' If we cannot find any zip files, return false

			ioFiles = ioDatasetFolder.GetFiles(ZIPPED_IMAGING_FILE_SEARCH_SPEC)
			If ioFiles Is Nothing OrElse ioFiles.Length = 0 Then
				' 0_R*.zip files not found
				MyBase.ReportError(ZIPPED_IMAGING_FILE_SEARCH_SPEC & "files not found in " & ioDatasetFolder.FullName)
				blnSuccess = False
			Else

				' Initialize the .DatasetFileInfo
				With mDatasetStatsSummarizer.DatasetFileInfo
					.FileSystemCreationTime = ioFiles(0).CreationTime
					.FileSystemModificationTime = ioFiles(0).LastWriteTime

					.AcqTimeStart = .FileSystemModificationTime
					.AcqTimeEnd = .FileSystemModificationTime

					.DatasetID = udtFileInfo.DatasetID
					.DatasetName = ioDatasetFolder.Name
					.FileExtension = ioFiles(0).Extension
					.FileSizeBytes = 0
					.ScanCount = 0
				End With


				' Update the dataset name and file extension
				udtFileInfo.DatasetName = GetDatasetNameViaPath(ioDatasetFolder.FullName)
				udtFileInfo.FileExtension = String.Empty

				udtFileInfo.AcqTimeEnd = System.DateTime.MinValue
				udtFileInfo.AcqTimeStart = System.DateTime.MaxValue
				udtFileInfo.ScanCount = 0

				' Process each zip file
				For Each ioFileInfo As System.IO.FileInfo In ioFiles

					' Examine all of the apexAcquisition.method files in this zip file
					blnSuccess = DetermineAcqStartEndTime(ioFileInfo, udtFileInfo)

				Next

				If udtFileInfo.AcqTimeEnd = System.DateTime.MinValue OrElse udtFileInfo.AcqTimeStart = System.DateTime.MaxValue Then
					' Did not find any apexAcquisition.method files or submethods.xml files
					' Use the file modification date of the first zip file
					udtFileInfo.AcqTimeStart = ioFiles(0).LastWriteTime
					udtFileInfo.AcqTimeEnd = ioFiles(0).LastWriteTime
				End If


				' Copy over the updated filetime info and scan info from udtFileInfo to mDatasetFileInfo
				With mDatasetStatsSummarizer.DatasetFileInfo
					.DatasetName = String.Copy(udtFileInfo.DatasetName)
					.FileExtension = String.Copy(udtFileInfo.FileExtension)
					.AcqTimeStart = udtFileInfo.AcqTimeStart
					.AcqTimeEnd = udtFileInfo.AcqTimeEnd
					.ScanCount = udtFileInfo.ScanCount
					.FileSizeBytes = udtFileInfo.FileSizeBytes
				End With

				blnSuccess = True
			End If
		Catch ex As System.Exception
			ReportError("Exception processing Zipped Imaging Files: " & ex.Message)
			blnSuccess = False
		End Try

		Return blnSuccess

	End Function


End Class

File: clsZippedImagingFilesScanner.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

// Written by Matthew Monroe for the Department of Energy (PNNL, Richland, WA)
//

namespace MSFileInfoScanner
{
    public class clsZippedImagingFilesScanner : clsMSFileInfoProcessorBaseClass
    {

        public const string ZIPPED_IMAGING_FILE_SEARCH_SPEC = "0_R*.zip";

        public const string ZIPPED_IMAGING_FILE_NAME_PREFIX = "0_R";

        /// <summary>
        /// Examines the subdirectories in the specified zip file
        /// Determines the oldest and newest modified analysis.baf files (or apexAcquisition.method file if analysis.baf files are not found)
        /// Presumes this is the AcqStartTime and AcqEndTime
        /// </summary>
        /// <param name="zipFile"></param>
        /// <param name="datasetFileInfo"></param>
        /// <returns>True if at least one valid file is found; otherwise false</returns>
        /// <remarks></remarks>
        private void DetermineAcqStartEndTime(FileInfo zipFile, clsDatasetFileInfo datasetFileInfo)
        {

            var success = false;

            try
            {
                // Bump up the file size
                datasetFileInfo.FileSizeBytes += zipFile.Length;

                var fileNamesToFind = new List<string> {
                    "analysis.baf",
                    "apexAcquisition.method",
                    "submethods.xml"};

                var oZipFile = new Ionic.Zip.ZipFile(zipFile.FullName);

                foreach (var fileNameToFind in fileNamesToFind)
                {
                    using (var zipEntry = oZipFile.GetEnumerator())
                    {

                        while (zipEntry.MoveNext())
                        {
                            if (zipEntry.Current == null)
                                continue;

                            if (zipEntry.Current.IsDirectory)
                            {
                                continue;
                            }

                            // Split the filename on the forward slash character
                            var nameParts = zipEntry.Current.FileName.Split('/');

                            if (nameParts.Length <= 0)
                            {
                                continue;
                            }

                            if (!string.Equals(nameParts[nameParts.Length - 1], fileNameToFind, StringComparison.CurrentCultureIgnoreCase))
                                continue;

                            if (zipEntry.Current.LastModified < datasetFileInfo.AcqTimeStart)
                            {
                                datasetFileInfo.AcqTimeStart = zipEntry.Current.LastModified;
                            }

                            if (zipEntry.Current.LastModified > datasetFileInfo.AcqTimeEnd)
                            {
                                datasetFileInfo.AcqTimeEnd = zipEntry.Current.LastModified;
                            }

                            // Bump up the scan count
                            datasetFileInfo.ScanCount += 1;

                            // Add a Scan Stats entry
                            var scanStatsEntry = new clsScanStatsEntry
                            {
                                ScanNumber = datasetFileInfo.ScanCount,
                                ScanType = 1,
                                ScanTypeName = "MALDI-HMS",
                                ScanFilterText = "",
                                ElutionTime = "0",
                                TotalIonIntensity = "0",
                                BasePeakIntensity = "0",
                                BasePeakMZ = "0",
                                BasePeakSignalToNoiseRatio = "0",
                                IonCount = 0,
                                IonCountRaw = 0
                            };

                            // Base peak signal to noise ratio

                            mDatasetStatsSummarizer.AddDatasetScan(scanStatsEntry);

                            success = true;
                        }
                    }

                    if (success)
                        break;
                }

            }
            catch (Exception ex)
            {
                OnErrorEvent("Error in DetermineAcqStartEndTime: " + ex.Message);
            }

        }

        private DirectoryInfo GetDatasetDirectory(string dataFilePath)
        {

            // First see if dataFilePath points to a valid file
            var datasetFile = new FileInfo(dataFilePath);

            if (datasetFile.Exists)
            {
                // User specified a file; assume the parent directory of this file is the dataset directory
                return datasetFile.Directory;
            }

            // Assume this is the path to the dataset directory
            return new DirectoryInfo(dataFilePath);
        }

        public override string GetDatasetNameViaPath(string dataFilePath)
        {
            var datasetName = string.Empty;

            try
            {
                // The dataset name for a dataset with zipped imaging files is the name of the parent directory
                // However, dataFilePath could be a file or a directory path, so use GetDatasetDirectory to get the dataset directory
                var datasetDirectory = GetDatasetDirectory(dataFilePath);
                datasetName = datasetDirectory.Name;

                if (datasetName.ToLower().EndsWith(".d"))
                {
                    datasetName = datasetName.Substring(0, datasetName.Length - 2);
                }

            }
            catch (Exception)
            {
                // Ignore errors
            }

            return datasetName;

        }

        public static bool IsZippedImagingFile(string imagingFilePath)
        {

            if (string.IsNullOrWhiteSpace(imagingFilePath))
            {
                return false;
            }

            var imagingFile = new FileInfo(imagingFilePath);

            if (imagingFile.Name.ToLower().StartsWith(ZIPPED_IMAGING_FILE_NAME_PREFIX.ToLower()) && imagingFile.Extension.ToLower() == ".zip")
            {
                return true;
            }

            return false;
        }

        /// <summary>
        /// Process the dataset
        /// </summary>
        /// <param name="dataFilePath"></param>
        /// <param name="datasetFileInfo"></param>
        /// <returns>True if success, False if an error</returns>
        /// <remarks></remarks>
        public override bool ProcessDataFile(string dataFilePath, clsDatasetFileInfo datasetFileInfo)
        {

            bool success;

            ResetResults();

            try
            {
                // Determine whether dataFilePath points to a file or a directory

                var datasetDirectory = GetDatasetDirectory(dataFilePath);

                // Validate that we have selected a valid directory
                if (!datasetDirectory.Exists)
                {
                    OnErrorEvent("File/directory not found: " + dataFilePath);
                    return false;
                }

                // In case we cannot find any .Zip files, update the .AcqTime values to the directory creation date
                datasetFileInfo.AcqTimeStart = datasetDirectory.CreationTime;
                datasetFileInfo.AcqTimeEnd = datasetDirectory.CreationTime;

                // Look for the 0_R*.zip files
                // If we cannot find any zip files, return false

                var zipFiles = datasetDirectory.GetFiles(ZIPPED_IMAGING_FILE_SEARCH_SPEC).ToList();
                if (zipFiles.Count == 0)
                {
                    // 0_R*.zip files not found
                    OnErrorEvent(ZIPPED_IMAGING_FILE_SEARCH_SPEC + "files not found in " + datasetDirectory.FullName);
                    success = false;

                }
                else
                {
                    var firstImagingFile = zipFiles.First();

                    // Initialize the .DatasetFileInfo
                    mDatasetStatsSummarizer.DatasetFileInfo.FileSystemCreationTime = firstImagingFile.CreationTime;
                    mDatasetStatsSummarizer.DatasetFileInfo.FileSystemModificationTime = firstImagingFile.LastWriteTime;

                    mDatasetStatsSummarizer.DatasetFileInfo.AcqTimeStart = mDatasetStatsSummarizer.DatasetFileInfo.FileSystemModificationTime;
                    mDatasetStatsSummarizer.DatasetFileInfo.AcqTimeEnd = mDatasetStatsSummarizer.DatasetFileInfo.FileSystemModificationTime;

                    mDatasetStatsSummarizer.DatasetFileInfo.DatasetID = datasetFileInfo.DatasetID;
                    mDatasetStatsSummarizer.DatasetFileInfo.DatasetName = datasetDirectory.Name;
                    mDatasetStatsSummarizer.DatasetFileInfo.FileExtension = firstImagingFile.Extension;
                    mDatasetStatsSummarizer.DatasetFileInfo.FileSizeBytes = 0;
                    mDatasetStatsSummarizer.DatasetFileInfo.ScanCount = 0;

                    // Update the dataset name and file extension
                    datasetFileInfo.DatasetName = GetDatasetNameViaPath(datasetDirectory.FullName);
                    datasetFileInfo.FileExtension = string.Empty;

                    datasetFileInfo.AcqTimeEnd = DateTime.MinValue;
                    datasetFileInfo.AcqTimeStart = DateTime.MaxValue;
                    datasetFileInfo.ScanCount = 0;

                    // Process each zip file

                    foreach (var zipFile in zipFiles)
                    {
                        // Examine all of the apexAcquisition.method files in this zip file
                        DetermineAcqStartEndTime(zipFile, datasetFileInfo);

                        if (mDisableInstrumentHash)
                        {
                            mDatasetStatsSummarizer.DatasetFileInfo.AddInstrumentFile(zipFile);
                        }
                        else
                        {
                            mDatasetStatsSummarizer.DatasetFileInfo.AddInstrumentFile(zipFile);
                        }
                    }

                    if (datasetFileInfo.AcqTimeEnd == DateTime.MinValue || datasetFileInfo.AcqTimeStart == DateTime.MaxValue)
                    {
                        // Did not find any apexAcquisition.method files or submethods.xml files
                        // Use the file modification date of the first zip file
                        datasetFileInfo.AcqTimeStart = firstImagingFile.LastWriteTime;
                        datasetFileInfo.AcqTimeEnd = firstImagingFile.LastWriteTime;
                    }

                    // Copy over the updated file time info and scan info from datasetFileInfo to mDatasetStatsSummarizer.DatasetFileInfo
                    mDatasetStatsSummarizer.DatasetFileInfo.DatasetName = string.Copy(datasetFileInfo.DatasetName);
                    mDatasetStatsSummarizer.DatasetFileInfo.FileExtension = string.Copy(datasetFileInfo.FileExtension);
                    mDatasetStatsSummarizer.DatasetFileInfo.AcqTimeStart = datasetFileInfo.AcqTimeStart;
                    mDatasetStatsSummarizer.DatasetFileInfo.AcqTimeEnd = datasetFileInfo.AcqTimeEnd;
                    mDatasetStatsSummarizer.DatasetFileInfo.ScanCount = datasetFileInfo.ScanCount;
                    mDatasetStatsSummarizer.DatasetFileInfo.FileSizeBytes = datasetFileInfo.FileSizeBytes;

                    success = true;
                }
            }
            catch (Exception ex)
            {
                OnErrorEvent("Exception processing Zipped Imaging Files: " + ex.Message);
                success = false;
            }

            PostProcessTasks();
            return success;

        }

    }
}